### PR TITLE
Add analyzer if Gateway not declared in the top-level gateways field of the VirtualService

### PR DESCRIPTION
**Pull request by kebe7jun**
For: https://github.com/istio/istio/issues/31916

[X] User Experience

### CODE REVIEW

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 1]
```diff
@@ -447,3 +447,16 @@ messages:
     args:
       - name: port
         type: string
+
+  - name: "GatewayNotDeclared"
+    code: IST0144
```
Comment (shamsher31) [inline galley/pkg/config/analysis/msg/messages.yaml:452]: I have used the same code in one of my [PR](https://github.com/istio/istio/pull/31993/files#diff-5ab445fd2493bb54c4667c7e74f7c0e5773ab35bf24fc5e639ab24d90a333274R452) which is still open. Probably I will have to use another code.

Comment (kebe7jun) [reply galley/pkg/config/analysis/msg/messages.yaml:452]: oh sorry， i will change it to another one.

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 2]
```diff
@@ -447,3 +447,16 @@ messages:
     args:
       - name: port
         type: string
+
+  - name: "GatewayNotDeclared"
+    code: IST0145
+    level: Warning
+    description: "Gateway in `match` is not declared in the gateways"
+    template: "Gateway %s which defined in .%s[].match[].gateways[] was not declared in .spec.gateways[] in virtual service %s"
```
Comment (shamsher31) [inline galley/pkg/config/analysis/msg/messages.yaml:455]:
```suggestion
    template: "Gateway %s which is defined in .%s[].match[].gateways[] is not declared in .spec.gateways[] in virtual service %s"
```

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 3]
```diff
@@ -447,3 +447,16 @@ messages:
     args:
       - name: port
         type: string
+
+  - name: "GatewayNotDeclared"
+    code: IST0145
+    level: Warning
+    description: "Gateway in `match` is not declared in the gateways"
```
Comment (shamsher31) [inline galley/pkg/config/analysis/msg/messages.yaml:454]: This does not sound descriptive. We should be more specific here.

**Review by esnible — CHANGES_REQUESTED** [review]
The original report https://github.com/istio/istio/issues/31916 says there is a false positive, presumably in `ReferencedResourceNotFound`.

The documentation https://istio.io/latest/docs/reference/config/networking/virtual-service/#HTTPMatchRequest says if there is a Gateways in the match requests these "Gateway names in the top-level gateways field of the VirtualService (if any) are overridden."

It may be poor style to define `.spec.gateways` then override them with `.spec.http.match.gateways`, but it isn't a warning.

To eliminate the false positive, we need to suppress the ReferencedResourceNotFound for `.spec.gateways` if EVERY match has a `gateways` field.

The current code does not look at `http.match.gateways`, `tcp.match.gateways`, or `tls.match.gateways`.  It SHOULD generate a ReferencedResourceNotFound if these refer to gateways that don't exist.

It might be worthwhile to generate an INFO if the `.spec.gateways` are all "dead" because there are match clauses that override them, as in @howardjohn 's example.  But if there is even one `http` with empty `match` or empty `match.gateways`, we shouldn't generate.

File: galley/pkg/config/analysis/analyzers/virtualservice/gateways.go [hunk 4]
```diff
@@ -85,6 +87,40 @@ func (s *GatewayAnalyzer) analyzeVirtualService(r *resource.Instance, c analysis
 			c.Report(collections.IstioNetworkingV1Alpha3Virtualservices.Name(), m)
 		}
 	}
+	gatewayNotDeclaredReport := func(g string, protocol string, ri, mi, gi int) {
+		m := msg.NewGatewayNotDeclared(r, g, protocol, vsName.String())
```
Comment (howardjohn) [inline galley/pkg/config/analysis/analyzers/virtualservice/gateways.go:91]: Should this just be a NewReferencedResourceNotFound like L71?

**Comment by kebe7jun** [comment]
not stale

**Comment by kebe7jun** [comment]
@esnible PTAL.

**Comment by kebe7jun** [comment]
not stale

**Comment by kebe7jun** [comment]
@esnible @howardjohn This PR is 3 months old, can you please take a look? Or maybe we don't need this feature for a while?

**Pull request closed without merging**